include/nil/crypto3/zk/snark/systems/plonk/placeholder/gates_argument.hpp

```diff
@@ -102,100 +71,139 @@ namespace nil {
+                                assignment = math::polynomial_shift(assignment, var.rotation, domain->m);
+                            }
+                            if (count > 1) {
+                                assignment.resize(extended_domain_size);
+                            }
+                            variable_values_out[var] = assignment;
+                        }
+                    }
+
                     static inline std::array<polynomial_dfs_type, argument_size>
                         prove_eval(
                             const typename policy_type::constraint_system_type &constraint_system,
                             const plonk_polynomial_dfs_table<FieldType, typename ParamsType::arithmetization_params>
                                 &column_polynomials,
-                            std::shared_ptr<math::evaluation_domain<FieldType>>
-                                original_domain,
+                            std::shared_ptr<math::evaluation_domain<FieldType>> original_domain,
                             std::uint32_t max_gates_degree,
                             transcript_type& transcript) {
                         PROFILE_PLACEHOLDER_SCOPE("gate_argument_time");
 
-                        std::uint32_t extended_domain_size = original_domain->m * 
-                            std::pow(2, ceil(std::log2(max_gates_degree)));
-                        
-                        const plonk_polynomial_dfs_table<FieldType, typename ParamsType::arithmetization_params>
-                            extended_column_polynomials = resize(column_polynomials, extended_domain_size);
-
+                        // max_gates_degree that comes from the outside does not take into account multiplication
+                        // by selector.
+                        ++max_gates_degree;
                         typename FieldType::value_type theta = transcript.template challenge<FieldType>();
 
-                        auto value_type_to_polynomial_dfs = [&assignments=extended_column_polynomials](
+                        auto value_type_to_polynomial_dfs = [](
                             const typename variable_type::assignment_type& coeff) {
                                 return polynomial_dfs_type(0, 1, coeff);
                             };
 
-                        bool is_final_expression_zero = true;
-                        math::expression<polynomial_dfs_variable_type> expr; 
+                        std::vector<std::uint32_t> extended_domain_sizes;
+                        std::vector<std::uint32_t> degree_limits;
+                        std::uint32_t max_degree = std::pow(2, ceil(std::log2(max_gates_degree)));
+                        std::uint32_t max_domain_size = original_domain->m * max_degree;
 
-                        typename FieldType::value_type theta_acc = FieldType::value_type::one();
+                        degree_limits.push_back(max_degree);

```

Comment:
Compute gate argument in 2 different sizes, whatever is less than max_degree / 2 goes into 1 expression, the rest goes into another. This way most of the gate argument is computed in half the size.